### PR TITLE
Fix stats api

### PR DESCRIPTION
**Pull request by lielShraga**

### CODE REVIEW

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 1]
```diff
@@ -982,8 +982,11 @@ class FTSService(config: Config) extends NsSplitter{
     def filterToBuilder(filter:AggregationFilter):AggregationBuilder = {
 
       implicit def fieldValueToValue(fieldValue: Field): String = fieldValue.operator match {
-        case AnalyzedField    => reverseNsTypedField(fieldValue.value)
-        case NonAnalyzedField => s"infoclone.${reverseNsTypedField(fieldValue.value)}.%exact"
+        case AnalyzedField =>  reverseNsTypedField(fieldValue.value)
+        case NonAnalyzedField=>
+          val fType = fieldType(fieldValue.value)
+          var reversed = reverseNsTypedField(fieldValue.value)
```
Comment (e-orz) [inline server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: use val

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 2]
```diff
@@ -982,8 +982,11 @@ class FTSService(config: Config) extends NsSplitter{
...
+          if (fType eq StringType) reversed + ".%exact" else reversed
```
Comment (e-orz) [inline server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: aggregation over system fields won't work. Should be fixed.
Check that is works with system fields for:
1. sort
1. search

File: server/cmwell-ws/app/controllers/Application.scala [hunk 3]
```diff
@@ -2066,7 +2066,9 @@ callback=< [URL] >
       new BadFieldTypeException(
         "Cannot cast field to numeric value. Did you try to use stats or histogram aggregations on non numeric field?",
         e
-      )
+      )case e:IllegalArgumentException
```
Comment (e-orz) [inline server/cmwell-ws/app/controllers/Application.scala]: style (\n)

File: server/cmwell-ws/app/controllers/Application.scala [hunk 4]
```diff
@@ -2066,7 +2066,9 @@ callback=< [URL] >
...
+      if e.getMessage.contains("aggregations failure due to fielddata disabled")
+      => new IllegalArgumentException("Stats API does not support non-exact value operator. Please use :: instead of :" + e.getMessage, e)
```
Comment (e-orz) [inline server/cmwell-ws/app/controllers/Application.scala]: please remove the "e".
Please elaborate that "non-exact value operator" is not supported only for string fields.

**Review by e-orz — CHANGES_REQUESTED** [review]

File: server/cmwell-ws/app/controllers/Application.scala (hunk 4 above)
Comment (lielShraga) [reply server/cmwell-ws/app/controllers/Application.scala]: done

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala (hunk 2 above)
Comment (lielShraga) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done for stats

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala (hunk 1 above)
Comment (lielShraga) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done

File: server/cmwell-ws/app/controllers/Application.scala (hunk 3 above)
Comment (lielShraga) [reply server/cmwell-ws/app/controllers/Application.scala]: done

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 5]
```diff
@@ -982,10 +990,17 @@ class FTSService(config: Config) extends NsSplitter{
     def filterToBuilder(filter:AggregationFilter):AggregationBuilder = {
 
       implicit def fieldValueToValue(fieldValue: Field): String = fieldValue.operator match {
-        case AnalyzedField =>  reverseNsTypedField(fieldValue.value)
+        case AnalyzedField =>
+          if(isStrSystemField(fieldValue.value))
+            throw new IllegalArgumentException("aggregations failure due to text system field")
+          val fType = fieldType(fieldValue.value)
+          if(fType eq StringType) {
```
Comment (e-orz) [inline server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: remove redundant braces

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 6]
```diff
@@ -300,6 +300,14 @@ class FTSService(config: Config) extends NsSplitter{
     }
   }
 
+  private def isStrSystemField(fieldName: String):Boolean = {
+    fieldName match {
+      case "system.parent" |"system.path" | "system.kind" | "system.uuid" | "system.dc" | "system.quad" |
```
Comment (e-orz) [inline server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: system.parent.parent_hierarchy is missing here

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 7]
```diff
@@ -982,10 +990,17 @@ class FTSService(config: Config) extends NsSplitter{
...
+        case AnalyzedField =>
+          if(isStrSystemField(fieldValue.value))
+            throw new IllegalArgumentException("aggregations failure due to text system field")
+          val fType = fieldType(fieldValue.value)
+          if(fType eq StringType) {
+            throw new IllegalArgumentException("aggregations failure due to fielddata disabled")
+          }
+          else reverseNsTypedField(fieldValue.value)
         case NonAnalyzedField=>
           val fType = fieldType(fieldValue.value)
```
Comment (e-orz) [inline server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: String system fields cannot be aggregated.

File: server/cmwell-ws/app/controllers/Application.scala [hunk 8]
```diff
@@ -2060,15 +2060,14 @@ callback=< [URL] >
...
       new BadFieldTypeException(
-        "Cannot cast field to numeric value. Did you try to use stats or histogram aggregations on non numeric field?",
-        e
-      )case e:IllegalArgumentException
-      if e.getMessage.contains("aggregations failure due to fielddata disabled")
-      => new IllegalArgumentException("Stats API does not support non-exact value operator. Please use :: instead of :" + e.getMessage, e)
+        "Cannot cast field to numeric value. Did you try to use stats or histogram aggregations on non numeric field?", e)
+    case e:IllegalArgumentException
+      if e.getMessage.contains("aggregations failure due to text system field")
+    => new IllegalArgumentException("Stats API does not support non-exact value operator for text system fields. Please use :: instead of :", e)
```
Comment (e-orz) [inline server/cmwell-ws/app/controllers/Application.scala]: at all

**Review by e-orz — CHANGES_REQUESTED** [review]

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala (hunk 6 above)
Comment (lielShraga) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala (hunk 7 above)
Comment (lielShraga) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 9]
```diff
@@ -989,19 +986,20 @@ class FTSService(config: Config) extends NsSplitter{
...
-        case NonAnalyzedField=>
-          val fType = fieldType(fieldValue.value)
-          val reversed = reverseNsTypedField(fieldValue.value)
-          if (fType eq StringType) reversed + ".%exact" else reversed
+      implicit def fieldValueToValue(fieldValue: Field): String = {
+        if (isStrSystemField(fieldValue.value, fieldType(fieldValue.value)))
+          throw new IllegalArgumentException("aggregations failure due to text system field")
+        fieldValue.operator match {
+          case AnalyzedField =>
+            val fType = fieldType(fieldValue.value)
```
Comment (e-orz) [inline server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: move this up to save calls

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala [hunk 10]
```diff
@@ -971,6 +963,11 @@ class FTSService(config: Config) extends NsSplitter{
 
   }
 
+  def isStrSystemField(fieldName: String, fieldType: FieldType with Product with Serializable): Boolean = {
```
Comment (e-orz) [inline server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: `with Product with Serializable` ???

**Review by e-orz — CHANGES_REQUESTED** [review]

File: server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala (hunk 10 above)
Comment (lielShraga) [reply server/cmwell-fts/src/main/scala/cmwell/fts/FTSService.scala]: done

File: server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+import scala.concurrent.duration._
+import scala.concurrent.{Await, Future}
+import scala.util.{Failure, Success, Try}
+
+class AggregationTests extends AsyncFunSpec with Matchers with Inspectors with Helpers with fixture.NSHashesAndPrefixes with LazyLogging {
+
+  describe("Agg API should") {
+    //Assertions
+    val ingestGeonames = Future.traverse(Seq(293846)) { n =>
```
Comment (e-orz) [inline server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala]: no need to to Future here.

File: server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+      val agg = scala.io.Source.fromURL(this.getClass.getResource(s"/agg/aggnames_$n.nq")).mkString
```
Comment (e-orz) [inline server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala]: resource leak! You should close the the resource. Please dismiss this remark if it is fixed in scala 2.12.

File: server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+import com.typesafe.scalalogging.LazyLogging
+import org.scalatest.{AsyncFunSpec, Inspectors, Matchers}
+import play.api.libs.json.{JsValue, _}
+
+import scala.concurrent.duration._
+import scala.concurrent.{Await, Future}
+import scala.util.{Failure, Success, Try}
+
+class AggregationTests extends AsyncFunSpec with Matchers with Inspectors with Helpers with fixture.NSHashesAndPrefixes with LazyLogging {
```
Comment (e-orz) [inline server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala]: There is no need for `fixture.NSHashesAndPrefixes`

File: server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+
+  describe("Agg API should") {
+    //Assertions
+    val ingestGeonames = Future.traverse(Seq(293846)) { n =>
+      val agg = scala.io.Source.fromURL(this.getClass.getResource(s"/agg/aggnames_$n.nq")).mkString
+      Http.post(_in, agg, Some("text/nquads;charset=UTF-8"), List("format" -> "nquads"), tokenHeader)
+    }.map { results =>
+      forAll(results) { res =>
```
Comment (e-orz) [inline server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala]: There is only one ingest - no need for `forAll`

File: server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala (hunk 11 above)
Comment (e-orz) [inline server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala]: rename all `geonames`

File: server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+      }
+    }
+
+
+    val path = cmw / "test.agg.org" / "Test201903_05_1501_11" / "testStatsApiTerms"
+
+    val aggForIntField = executeAfterCompletion(ingestGeonames) {
+      spinCheck(100.millis, true)(Http.get(
+        uri = path,
+        queryParams = List("op" -> "stats", "format" -> "json", "debug-info" -> "", "ap" -> "type:term,field::num.testns,size:3"))) { r =>
```
Comment (e-orz) [inline server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala]: you should use the full URI as preparation for distributed system tests
e.g. `num.$http://qa.test.rfnt.com/v1.1/testns$`

File: server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+        (Json.parse(r.payload) \ "AggregationResponse" \\ "buckets": @unchecked) match {
+          case n: Seq[JsValue] => (r.status == 200) && n.forall(jsonval=> jsonval.as[JsArray].value.size == 3)
+        }
+      }.map { res =>
+        withClue(res) {
+          res.status should be(200)
+          val total = (Json.parse(res.payload) \ "AggregationResponse" \\ "buckets").map(jsonval=> jsonval.as[JsArray].value.size)
+          total should contain(3)
```
Comment (e-orz) [inline server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala]: should be(3)

File: server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+        }
+      }
+    }
+
+
+    val aggForExactTextField = executeAfterCompletion(ingestGeonames) {
+      spinCheck(100.millis, true)(Http.get(
+        uri = path,
+        queryParams = List("op" -> "stats", "format" -> "json", "debug-info" -> "", "ap" -> "type:term,field::Test_Data.testns,size:2"))) { r =>
```
Comment (e-orz) [inline server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala]: testns

**Review by e-orz — CHANGES_REQUESTED** [review]

File: server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala (hunk 11 above)
Comment (lielShraga) [reply server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala]: done

File: server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala (hunk 13 above)
Comment (lielShraga) [reply server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala]: done

File: server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala (hunk 14 above)
Comment (lielShraga) [reply server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala]: done

File: server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala (hunk 11 above)
Comment (lielShraga) [reply server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala]: done

File: server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala [hunk 18]
```diff
@@ -16,71 +16,75 @@
...
     }
 
+    agg.close()
+
 
     val path = cmw / "test.agg.org" / "Test201903_05_1501_11" / "testStatsApiTerms"
 
-    val aggForIntField = executeAfterCompletion(ingestGeonames) {
+    val aggForIntField = executeAfterCompletion(ingestAgg) {
+      val fieldVal = URLEncoder.encode("$http://qa.test.rfnt.com/v1.1/testns#num$")
```
Comment (e-orz) [inline server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala]: remove this line

File: server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala [hunk 19]
```diff
@@ -16,71 +16,75 @@
 
 package cmwell.it
 
+import java.net.URLEncoder
```
Comment (e-orz) [inline server/cmwell-it/src/it/scala/cmwell/it/AggregationTests.scala]: remove this line

**Review by e-orz — CHANGES_REQUESTED** [review]

**Pull request merged**